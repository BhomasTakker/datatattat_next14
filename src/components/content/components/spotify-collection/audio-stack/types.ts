import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
