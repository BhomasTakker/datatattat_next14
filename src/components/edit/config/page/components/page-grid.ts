import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const PAGE_GRID_CONFIG: InputListProps = {
	id: "grid",
	type: EditInputs.inputList,
	label: "Page Grid",

	inputs: [
		{
			id: "variant",
			type: EditInputs.select,
			label: "Grid Variant",
			options: [""],
		},
	],
};
