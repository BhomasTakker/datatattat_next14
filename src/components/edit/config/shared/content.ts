import { SelectInputProps } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../inputs/inputs";
import {
	ContainerHeightOptions,
	ContainerWidthOptions,
} from "@/components/page/components/stack/types";

const filterOptions = (includes: string[], options: Record<string, string>) => {
	return Object.values(options).filter((option) => includes.includes(option));
};

export const useContainerWidth = (
	id: string,
	label: string,
	includes: string[] = Object.values(ContainerWidthOptions)
): SelectInputProps => {
	return {
		id,
		type: EditInputs.select,
		label,
		defaultValue: ContainerWidthOptions.MD,
		options: filterOptions(includes, ContainerWidthOptions),
	};
};

export const useContainerHeight = (
	id: string,
	label: string,
	includes: string[] = Object.values(ContainerHeightOptions)
): SelectInputProps => {
	return {
		id,
		type: EditInputs.select,
		label,
		defaultValue: ContainerHeightOptions.MD,
		options: filterOptions(includes, ContainerHeightOptions),
	};
};
