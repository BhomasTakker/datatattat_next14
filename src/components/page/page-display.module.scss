@use "../../scss/abstracts/units/" as *;

.page {
	display: flex;
	flex-direction: column;
}
