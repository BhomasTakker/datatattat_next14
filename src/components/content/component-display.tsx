import { PageComponent } from "@/types/page";
import { ComponentProfile } from "./component-profile/component-profile";
import { ComponentFactory } from "./components/component-factory";

export const ComponentDisplay = ({
	component,
}: {
	component: PageComponent;
}) => {
	return (
		<section data-testid="component">
			<ComponentProfile profile={{}} />
			<ComponentFactory component={component} />
		</section>
	);
};
