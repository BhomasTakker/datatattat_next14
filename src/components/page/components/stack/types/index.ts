import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
