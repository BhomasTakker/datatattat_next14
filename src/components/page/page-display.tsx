import { IPage } from "@/types/page";
import { PageProfile } from "./profile/page-profile";
import { PageComponentFactory } from "./components/page-component-factory";
import styles from "./page-display.module.scss";
import { getCurrentRoute } from "@/utils/route";

export const PageDisplay = async ({ page }: { page: IPage }) => {
	const { content, meta, profile } = page || {};

	// not here
	const currentRoute = await getCurrentRoute();
	const isUserRoute = currentRoute?.includes("/users/") || false;
	const isClient = isUserRoute;

	return (
		<main className={styles.page}>
			{/* REPLACE with meta function */}
			{/* Page head for meta data, seo, etc */}
			{/* <PageHead headData={meta} /> */}
			{/* Page profile data for title, description, createdBy, etc */}
			<PageProfile profile={profile} />
			<PageComponentFactory content={content} isClient={isClient} />
		</main>
	);
};
