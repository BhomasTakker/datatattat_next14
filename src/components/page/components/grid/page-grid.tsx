import { PageComponents, PageContainer, PageContent } from "@/types/page";
import styles from "./page-grid.module.scss";
import { ComponentDisplay } from "@/components/content/component-display";

// We should probably have 'style' in the Props type
type PageGridProps = {
	layout: string;
};

export type PageGridContent = {
	containerType: "Grid";
	props: PageGridProps;
	components: PageComponents;
};

export const PageGrid = ({ content }: { content: PageContent }) => {
	const { props, components } = content;
	const { layout: style } = (props as PageGridProps) || {};

	/////////////////////////////////////////////////////////////////////
	// potentially pre load some data here i.e. first three components
	// the rest pass a function/object to load the data client side
	// then can lazy load the rest of the components
	/////////////////////////////////////////////////////////////////////
	const renderComponents = () => {
		return components.map((component, index) => {
			// potentially wrap in a div - if we need to control grid items
			// we should be able to use a pure css way?
			return <ComponentDisplay key={index} component={component} />;
		});
	};

	const className = `${styles.root} ${styles[`root-${style}`]}`;

	return (
		<section data-testid="page-grid" className={className}>
			{renderComponents()}
		</section>
	);
};
