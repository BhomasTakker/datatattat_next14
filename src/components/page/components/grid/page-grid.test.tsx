import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { PageGrid, PageGridContent } from "./page-grid";
import {} from "../../../content/component-display";
import { PageComponent } from "@/types/page";

jest.mock("../../../content/component-display", () => {
	return {
		__esModule: true,
		ComponentDisplay: ({ component }: { component: PageComponent }) => {
			const { componentType, componentProps } = component;
			return (
				<section data-testid="page-grid-item">
					{`Component type:- ${componentType}`}
					{`Component props:- ${JSON.stringify(componentProps)}`}
				</section>
			);
		},
	};
});

// we need some proper mock data for tests
const MOCK: PageGridContent = {
	containerType: "Grid",
	props: {
		layout: "test-grid",
	},
	components: [],
};

const MOCK_COMPONENT = {
	componentType: "GridItem",
	componentProps: {},
	_with: {},
};

describe("Page Grid Test Suite", () => {
	it("should render a page grid element", () => {
		render(<PageGrid content={MOCK} />);
		const component = screen.getByTestId("page-component");
		expect(component).toBeInTheDocument();
	});

	it("should render an div with expected class", () => {
		render(<PageGrid content={MOCK} />);
		const component = screen.getByTestId("page-component");
		expect(component).toHaveClass("root", "root-test-grid");
	});

	it("should render a grid item", () => {
		// @ts-expect-error - mocking data
		MOCK.components = [MOCK_COMPONENT];
		render(<PageGrid content={MOCK} />);
		const component = screen.getByTestId("page-grid-item");
		expect(component).toBeInTheDocument();
	});

	it("should render expected number of list items", () => {
		// @ts-expect-error - mocking data
		MOCK.components = [MOCK_COMPONENT, MOCK_COMPONENT, MOCK_COMPONENT];
		render(<PageGrid content={MOCK} />);
		const components = screen.getAllByTestId("page-grid-item");
		expect(components).toHaveLength(3);
	});

	it("renders PageGrid unchanged", () => {
		// @ts-expect-error - mocking data
		MOCK.components = [MOCK_COMPONENT];
		const { container } = render(<PageGrid content={MOCK} />);
		expect(container).toMatchSnapshot();
	});
});
