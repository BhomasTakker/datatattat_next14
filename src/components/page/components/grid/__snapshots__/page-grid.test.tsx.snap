// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Grid Test Suite renders PageGrid unchanged 1`] = `
<div>
  <section
    class="root root-test-grid"
    data-testid="page-component"
  >
    <div
      data-testid="content-component"
    >
      <section
        data-testid="page-grid-item"
      >
        Component type:- GridItem
        Component props:- {}
      </section>
    </div>
  </section>
</div>
`;
